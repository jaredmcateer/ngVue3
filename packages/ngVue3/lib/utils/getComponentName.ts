/**
 * Tries to get the component name or infer it from the file path if possible,
 * failing those returns a generic name
 * @param component Component to get name for
 * @returns a name
 */
export function getComponentName(component: any) {
  let name = "AnonymousComponent";

  if (component.name) {
    name = component.name;
  } else if (component.__file) {
    name = parseFileName(component.__file) || name;
  }

  return name;
}

/**
 * Converts the file path to a component name if possible, this uses the same
 * regex as Vue 3 does
 * @see https://github.com/vuejs/core/blob/8dcb6c7bbdd2905469e2bb11dfff27b58cc784b2/packages/runtime-core/src/component.ts#L966
 *
 * @param path file path of component
 * @returns An empty string if path doesn't match or the CamelCased name of
 *          component inferred from path
 */
function parseFileName(path: string) {
  let name = "";
  const match = path.match(/([^/\\]+)\.\w+$/);
  if (match) {
    name = match[1].replace(/(?:^|[-_])(\w)/g, (c) => c.toUpperCase()).replace(/[-_]/g, "");
  }
  return name;
}
