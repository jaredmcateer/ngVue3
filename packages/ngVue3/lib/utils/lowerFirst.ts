/**
 * Lowers the first character of the provided text. Note this function does not
 * handle locales as this is for HTML attribute names
 *
 * @param text
 * @returns The provided text with the first character lowered
 */
export function lowerFirst([first, ...rest]: string): string {
  return first.toLowerCase() + rest.join("");
}
