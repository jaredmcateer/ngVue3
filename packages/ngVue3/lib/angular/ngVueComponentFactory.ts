import angular from "angular";
import { ngVueLinker } from "./ngVueLinker";

/**
 * @example
 * angular.module('my.vue.components')
 *   .directive(...createVueComponent('HelloComponent', {
 *     props: {
 *       firstName: String,
 *       lastName: String,
 *     },
 *     emits: ['update-name'],
 *     setup(props, context) {
 *       const updateName = () => emit('update-name', 'Barb');
 *
 *       return () => (
 *         <h1>{props.firstName} {props.lastName}</h1>
 *         <button v-on:click={updateName}>Click me</button>
 *       );
 *     }
 *   }))
 *
 * @example
 * <hello-component
 *   v-props="ctrl.person"
 *   v-on-update-name="ctrl.nameUpdated"
 * ></hello-component>
 *
 * @example
 * <hello-component
 *   v-props-first-name="ctrl.person.firstName"
 *   v-props-last-name="ctrl.person.lastName"
 *   v-on-update-name="ctrl.nameUpdated"
 * ></hello-component>
 *
 * @param $injector
 * @returns A function when called returns a directive config
 */
export function ngVueComponentFactory($injector: ng.auto.IInjectorService) {
  return ngVueComponentDirective;
}
ngVueComponentFactory.$inject = ["$injector"];

export function ngVueComponentDirective(
  component: unknown,
  ngDirectiveConfig?: ng.IDirective
): ng.IDirective {
  const config: ng.IDirective = {
    restrict: "E",
    link(scope: ng.IScope, elem: JQLite, attrs: ng.IAttributes) {
      ngVueLinker(component, elem, attrs, scope);
    },
  };

  return angular.extend(config, ngDirectiveConfig);
}
