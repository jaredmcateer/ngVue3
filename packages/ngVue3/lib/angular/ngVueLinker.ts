import angular from "angular";
import { createApp, DefineComponent, h, onMounted, ref, toRefs } from "vue";
import { evaluateEvents } from "../components/evaluateEvents";
import { evaluateValues } from "../components/evaluateValues";
import { extractSpecialAttributes } from "../components/extractSpecialAttributes";
import { getExpressions } from "../components/getExpressions";
import { WatchExpressionOptions, watchExpressions } from "../components/watchExpressions";
import { watchSpecialAttributes } from "../components/watchSpecialAttributes";
import { getInstanceState } from "../instanceStore";

export function ngVueLinker(
  Component: any,
  jqElement: JQLite,
  attrs: ng.IAttributes,
  scope: ng.IScope
) {
  // Create unique key for retrieving state information
  const instanceKey = Symbol("ngVueInstanceKey");
  const state = getInstanceState(instanceKey);
  const removeAttrFn = getAttrRemoveFunction(jqElement, attrs);
  const attrExpressions = getExpressions(attrs, removeAttrFn);
  const events = evaluateEvents(attrExpressions.events, scope);

  Object.assign(state.props, evaluateValues(attrExpressions.props, scope));
  Object.assign(state.attrs, evaluateValues(attrExpressions.attrs, scope));
  Object.assign(state.special, extractSpecialAttributes(attrs, removeAttrFn));

  const options: WatchExpressionOptions = { depth: attrs.watchDepth };
  removeAttrFn("watchDepth");

  watchExpressions(attrExpressions.props, state.props, scope, options);
  watchExpressions(attrExpressions.attrs, state.attrs, scope, options);
  watchSpecialAttributes(state.special, jqElement, scope);

  let html = getInnerHtml(jqElement[0]);

  let vueInstance = createApp({
    name: `NgVue-${Component.name || "UnnamedComponent"}`,
    setup() {
      const slot = ref<HTMLSpanElement | null>(null);

      onMounted(() => {
        if (html) {
          slot.value?.replaceChild(html, slot.value);
        }
      });

      return () =>
        h(Component, { ...state.props, ...state.attrs, ...state.special, ...events }, [
          h("span", { ref: slot }),
        ]);
    },
  });

  vueInstance.mount(jqElement[0]);

  scope.$on("$destroy", () => {
    vueInstance.unmount();
    if (vueInstance._container) {
      angular.element(vueInstance._container).remove();
    }
    // @ts-ignore We're dereferencing this variable for garbage collection
    vueInstance = null;
  });
}

function getInnerHtml(element: HTMLElement) {
  if (element.innerHTML.trim()) {
    let html: Element;
    if (element.children.length === 0) {
      const span = document.createElement("span");
      span.innerHTML = element.innerHTML.trim();
      html = span;
    } else {
      html = element.children[0];
    }

    return html;
  }
}

function getAttrRemoveFunction(element: JQLite, attributes: ng.IAttributes) {
  return (attr: string) => {
    const denormalizedAttr = (attributes.$attr as Record<string, string>)[attr];
    element.removeAttr(denormalizedAttr);
  };
}
