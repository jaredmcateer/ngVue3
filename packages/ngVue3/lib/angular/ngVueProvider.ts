import angular from "angular";
import { App, Directive, InjectionKey, Plugin, Component } from "vue";

export type PluginHook = ($injector: ng.auto.IInjectorService, app: App<Element>) => void;

export interface NgVuePlugin {
  /** Plugin name */
  $name: string;
  /** Plugin install function */
  $plugin: PluginHook;
  /** Methods/properties to expose on the angular side as part of the ngVueProvider.plugins.<$name>.<$config.key> */
  $config: Record<string, unknown>;
}

export type NgVueInjectable = [key: string | InjectionKey<unknown>, value: unknown];
export type NgVueHookCallback = (hook: PluginHook) => void;

export interface NgVueService {
  initNgVuePlugins(app: App<Element>): void;
  getInjectables(): NgVueInjectable[];
  getVuePlugins(): Plugin[];
  getVueDirectives(): Record<string, Directive>;
  getVueComponents(): Record<string, Component>;
}

export class NgVueProvider {
  $get: ["$injector", ($injector: ng.auto.IInjectorService) => NgVueService];
  static readonly $inject = ["$injector"];

  private pluginHooks: PluginHook[] = [];
  private pluginConfig: Record<string, any> = {};
  private injectables: NgVueInjectable[] = [];
  private nativeVuePlugins: Plugin[] = [];
  private nativeVueDirectives: Record<string, Directive> = {};
  private nativeVueComponents: Record<string, Component> = {};

  constructor(private $injector: ng.auto.IInjectorService) {
    this.$get = [
      "$injector",
      ($injector: ng.auto.IInjectorService): NgVueService => {
        const initNgVuePlugins = (app: App<Element>) => {
          const callback: NgVueHookCallback = (hook) => {
            hook($injector, app);
          };

          this.callHooks(this.pluginHooks, callback);
        };

        return {
          initNgVuePlugins,
          getInjectables: () => this.injectables,
          getVuePlugins: () => this.nativeVuePlugins,
          getVueDirectives: () => this.nativeVueDirectives,
          getVueComponents: () => this.nativeVueComponents,
        };
      },
    ];
  }

  get plugins() {
    return this.pluginConfig;
  }

  /**
   * Acts as a pass through to the Vue app instance for providing an injectable.
   * @param name A string or symbol that can be used to inject into a component
   * @param value
   */
  provide(name: string, value: unknown): void;
  provide<IKey = unknown>(name: InjectionKey<IKey>, value: unknown): void;
  provide<IKey = unknown>(name: string | InjectionKey<IKey>, value: unknown) {
    this.injectables.push([name, value]);
  }

  /**
   * Acts as a pass through for native Vue plugins to the app instance. If
   * access to angular is needed by the plugin use `installNgVuePlugin`
   * @param vuePlugin Native Vue plugin
   */
  use(vuePlugin: Plugin) {
    this.nativeVuePlugins.push(vuePlugin);
  }

  /**
   * Acts as a pass through for native vue directives to the app instance.
   * @param vueDirective
   */
  directive(name: string, vueDirective: Directive) {
    this.nativeVueDirectives[name] = vueDirective;
  }

  /**
   * Acts as a pass through for native vue components to the app instance.
   * @param vueComponent
   */
  component(name: string, vueComponent: Component) {
    this.nativeVueComponents[name] = vueComponent;
  }

  /**
   * Installs an ngVue plugin, this gives access to configuration via the
   * ngVueProvider and gives the Vue plugin install method access to the angular
   * injector.
   *
   * @param plugin a function that returns a ngVue plugin config
   */
  installNgVuePlugin(plugin: () => NgVuePlugin) {
    const { $name, $config, $plugin } = plugin();
    this.pluginHooks.push($plugin);
    this.pluginConfig[$name] = $config;
  }

  private callHooks(hooks: PluginHook[], callback: NgVueHookCallback) {
    if (hooks) {
      hooks.forEach((hook) => callback(hook));
    }
  }
}

let ngVuePluginsModule: ng.IModule;

export function useNgVuePlugins() {
  if (!ngVuePluginsModule) {
    ngVuePluginsModule = angular.module("ngVue.plugins", []).provider("$ngVue", [
      "$injector",
      function ($injector: ng.auto.IInjectorService) {
        return new NgVueProvider($injector);
      },
    ]);
  }

  return ngVuePluginsModule.name;
}
