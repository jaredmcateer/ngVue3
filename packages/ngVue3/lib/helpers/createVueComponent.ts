import { ngVueComponentDirective } from "../angular/ngVueComponentFactory";

type NgDirectiveTuple<T extends string = string> = [
  T,
  ["createVueComponent", (createVueComponent: typeof ngVueComponentDirective) => ng.IDirective]
];
/**
 * Reduces boilerplate of creating vue components
 *
 * @example
 * angular
 *   .module(...)
 *   .directive(...createVueComponent('my-component', Component))
 *
 * Is the equivalent to
 *
 * angular
 *   .module(...)
 *   .directive(
 *     'my-component',
 *     [
 *       'createVueComponent',
 *       (createVueComponent) => createVueComponent(
 *         'my-component',
 *         Component
 *       )
 *     ]
 *   )
 *
 * @param name Name of ng directive
 * @param Component Vue Component
 * @param ngDirectiveConfig extends the default angular directive
 * @returns array to be used in directive declaration
 */
export function createVueComponent<N extends string>(
  name: N,
  Component: unknown,
  ngDirectiveConfig?: ng.IDirective
): NgDirectiveTuple<N> {
  return [
    name,
    [
      "createVueComponent",
      (createVueComponent: typeof ngVueComponentDirective) =>
        createVueComponent(Component, ngDirectiveConfig),
    ],
  ];
}
