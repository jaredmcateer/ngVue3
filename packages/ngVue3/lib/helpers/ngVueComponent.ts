import { NgVueComponentDirective } from "../angular/ngVueComponentFactory";

type NgDirectiveTuple<T extends string = string> = [
  T,
  ["createVueComponent", (createVueComponent: NgVueComponentDirective) => ng.IDirective]
];
/**
 * Reduces boilerplate of creating vue components
 *
 * @example
 * angular
 *   .module(...)
 *   .directive(...ngVueComponent('my-component', Component))
 *
 * Is the equivalent to
 *
 * angular
 *   .module(...)
 *   .directive(
 *     'my-component',
 *     [
 *       'createVueComponent',
 *       (createVueComponent) => createVueComponent(
 *         'my-component',
 *         Component
 *       )
 *     ]
 *   )
 *
 * @param name Name of ng directive
 * @param Component Vue Component
 * @param ngDirectiveConfig extends the default angular directive
 * @returns array to be used in directive declaration
 */
export function ngVueComponent<N extends string>(
  name: N,
  Component: unknown,
  ngDirectiveConfig?: ng.IDirective
): NgDirectiveTuple<N> {
  return [
    name,
    [
      "createVueComponent",
      (createVueComponent: NgVueComponentDirective) =>
        createVueComponent(Component, ngDirectiveConfig),
    ],
  ];
}
