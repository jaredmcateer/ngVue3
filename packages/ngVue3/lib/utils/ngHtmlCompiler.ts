import angular from "angular";

export function ngHtmlCompiler($compile: ng.ICompileService) {
  return (html: string, scope: ng.IScope): JQLite => {
    const element = angular.element(html);
    $compile(element)(scope);
    scope.$digest();
    return element;
  };
}
