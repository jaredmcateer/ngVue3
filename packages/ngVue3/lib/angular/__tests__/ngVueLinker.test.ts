import angular from "angular";
import ScriptSetupVue from "./__fixtures__/ScriptSetup.vue";
import CompositionComponentVue from "./__fixtures__/CompositionComponent.vue";
import OptionsComponent from "./__fixtures__/OptionsComponent.vue";
import { createVueComponent } from "../../helpers/createVueComponent";
import { nextTick } from "vue";
import { ngHtmlCompiler } from "../../utils/ngHtmlCompiler";

interface TestScope extends ng.IScope {
  user: { firstName: string; lastName: string };
  isAdmin: boolean;
}

describe.each`
  style                 | Component
  ${"Options API"}      | ${OptionsComponent}
  ${"Composition API"}  | ${CompositionComponentVue}
  ${"Script Setup API"} | ${ScriptSetupVue}
`("ngVueLinker ($style)", ({ Component }) => {
  let $compileProvider: ng.ICompileProvider;
  let scope: TestScope;
  let compile: ReturnType<typeof ngHtmlCompiler>;

  beforeEach(angular.mock.module("ngVue"));
  beforeEach(
    angular.mock.module((_$compileProvider_: ng.ICompileProvider) => {
      $compileProvider = _$compileProvider_;
    })
  );

  beforeEach(
    angular.mock.inject((_$rootScope_: ng.IRootScopeService, _$compile_: ng.ICompileService) => {
      compile = ngHtmlCompiler(_$compile_);
      scope = (function (): TestScope {
        const tmpScope: any = _$rootScope_.$new();
        tmpScope.user = { firstName: "Burt", lastName: "Macklin" };
        tmpScope.isAdmin = true;
        return tmpScope;
      })();
    })
  );

  beforeEach(() => {
    $compileProvider.directive(...createVueComponent("myComponent", Component));
  });

  describe("Render", () => {
    it("should render the component", () => {
      const element = compile("<my-component />", scope);
      expect(element[0]).toMatchSnapshot();
    });

    it("should not render the component if angular is preventing it", () => {
      const element = compile(`<my-component ng-if="false" />`, scope);
      expect(element[0]).toMatchSnapshot();
    });

    it("should render user details when provided as an object", () => {
      const element = compile(`<my-component v-props="user" />`, scope);
      expect(element[0]).toMatchSnapshot();
    });

    it("should render button when provided as a named prop", () => {
      const element = compile(`<my-component v-props-is-admin="isAdmin" />`, scope);
      expect(element[0]).toMatchSnapshot();
    });

    it("should render user details and button when provided as a mixed prop types", () => {
      const element = compile(`<my-component v-props="user" v-props-is-admin="isAdmin" />`, scope);
      expect(element[0]).toMatchSnapshot();
    });
  });

  describe("Property Reactivity", () => {
    describe(`(v-props-*="...")`, () => {
      it("should update named props when changes are made", async () => {
        const element = compile(
          `<my-component v-props-first-name="user.firstName" v-props-last-name="user.lastName" v-props-is-admin="isAdmin" />`,
          scope
        );
        scope.user.firstName = "Barb";
        scope.user.lastName = "Ara";
        scope.isAdmin = false;
        scope.$digest();
        await nextTick();

        expect(element[0]).toMatchSnapshot();
      });
    });

    describe(`(v-props="...")`, () => {
      it("should update when reference changes", async () => {
        const element = compile(
          `<my-component v-props="user" v-props-is-admin="isAdmin" />`,
          scope
        );
        scope.user = { firstName: "Bart", lastName: "Moklen" };
        scope.$digest();
        await nextTick();

        expect(element[0]).toMatchSnapshot();
      });

      it("should not update when property value changes when observing for reference changes", async () => {
        const element = compile(
          `<my-component v-props="user" v-props-is-admin="isAdmin" />`,
          scope
        );
        scope.user.firstName = "Bort";
        scope.user.lastName = "Muklyn";
        scope.$digest();
        await nextTick();

        expect(element[0]).toMatchSnapshot();
      });

      it("should update when property value changes only when observing at value depth", async () => {
        const element = compile(
          `<my-component v-props="user" v-props-is-admin="isAdmin" watch-depth="value" />`,
          scope
        );
        scope.user.firstName = "Bort";
        scope.user.lastName = "Muklyn";
        scope.$digest();
        await nextTick();

        expect(element[0]).toMatchSnapshot();
      });
    });
  });
});
