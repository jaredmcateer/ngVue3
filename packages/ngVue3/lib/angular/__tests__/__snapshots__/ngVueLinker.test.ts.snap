// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ngVueLinker (Composition API) Property Reactivity (v-props="...") should not update when property value changes when observing for reference changes 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Burt Macklin
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Property Reactivity (v-props="...") should update when property value changes only when observing at value depth 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Bort Muklyn
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Property Reactivity (v-props="...") should update when reference changes 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Bart Moklen
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Property Reactivity (v-props-*="...") should update named props when changes are made 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Barb Ara
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Render should not render the component if angular is preventing it 1`] = `<!-- ngIf: false -->`;

exports[`ngVueLinker (Composition API) Render should render button when provided as a named prop 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Not Sure
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Render should render the component 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Not Sure
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Render should render user details and button when provided as a mixed prop types 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Burt Macklin
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Composition API) Render should render user details when provided as an object 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Burt Macklin
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Options API) Property Reactivity (v-props="...") should not update when property value changes when observing for reference changes 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Burt Macklin
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Options API) Property Reactivity (v-props="...") should update when property value changes only when observing at value depth 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Bort Muklyn
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Options API) Property Reactivity (v-props="...") should update when reference changes 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Bart Moklen
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Options API) Property Reactivity (v-props-*="...") should update named props when changes are made 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Barb Ara
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Options API) Render should not render the component if angular is preventing it 1`] = `<!-- ngIf: false -->`;

exports[`ngVueLinker (Options API) Render should render button when provided as a named prop 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Not Sure
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Options API) Render should render the component 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Not Sure
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Options API) Render should render user details and button when provided as a mixed prop types 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Burt Macklin
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Options API) Render should render user details when provided as an object 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Burt Macklin
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Property Reactivity (v-props="...") should not update when property value changes when observing for reference changes 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Burt Macklin
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Property Reactivity (v-props="...") should update when property value changes only when observing at value depth 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Bort Muklyn
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Property Reactivity (v-props="...") should update when reference changes 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Bart Moklen
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Property Reactivity (v-props-*="...") should update named props when changes are made 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Barb Ara
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Render should not render the component if angular is preventing it 1`] = `<!-- ngIf: false -->`;

exports[`ngVueLinker (Script Setup API) Render should render button when provided as a named prop 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Not Sure
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Render should render the component 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Not Sure
  </div>
  <!--v-if-->
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Render should render user details and button when provided as a mixed prop types 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Hello, Burt Macklin
  </div>
  <button>
    Click Me
  </button>
  
</my-component>
`;

exports[`ngVueLinker (Script Setup API) Render should render user details when provided as an object 1`] = `
<my-component
  class="ng-scope"
  data-v-app=""
>
  
  <div>
    Welcome, Burt Macklin
  </div>
  <!--v-if-->
  
</my-component>
`;
